calyx/src/ir/component.rs

```diff
@@ -102,4 +97,56 @@ impl Component {
     {
         self.namegen.gen_name(prefix)
     }
+
+    /// Returns an iterator over the groups contained in this component.
+    pub fn iter_groups(&self) -> impl Iterator<Item = &RRC<Group>> {
+        self.groups.values()
+    }
+
+    /// Returns an iterator over the cells contained in this component.
+    pub fn iter_cells(&self) -> impl Iterator<Item = &RRC<Cell>> {
+        self.cells.values()
+    }
+
+    /// Adds the given group to the component.
+    pub fn add_group(&mut self, group: RRC<Group>) {
+        let name = group.borrow().name().clone();
+        self.groups.insert(name, group);
+    }
+
+    /// Adds the given cell to the component.
+    pub fn add_cell(&mut self, cell: RRC<Cell>) {
+        let name = cell.borrow().name().clone();
+        self.cells.insert(name, cell);
+    }
+    /// Iterates through the given vec and adds all the cells to the component.
+    pub fn add_cells(&mut self, cells: Vec<RRC<Cell>>) {
```

Comment:
I assume this can't be a slice of `RRC<Cell>` because slices only return things by reference?

Reply:
Yeah, since it's adding the cells it needs to have ownership over them. If we wanted the method to clone the RRCs it could take a slice. Another option would be to have it take something that implements Iterator over the owned values, rather than a reference to them. That might be a better choice since it doesn't really need to be a vector.

Reply:
is there a reason not to clone the Rc's here?

Reply:
The only reason not to is that the one place this functionality is needed explicitly constructs the vector to be passed in and so already does the requisite cloning. I don't think that's particularly compelling, but I was trying to make the minimum number of changes to the code behavior and tried to match the ownership behavior to the extent possible.